module/platform/Webots/src/Webots.cpp

```diff
@@ -0,0 +1,228 @@
+
+    // Store the ip address information that we will connect to
+    sockaddr_in address;
+    memset(&address, 0, sizeof(sockaddr_in));
+    address.sin_family = AF_INET;
+    address.sin_port   = htons(port);
+    hostent* server    = gethostbyname(server_name.c_str());
+
+    // Check that dns was successful
+    if (server) {
+        std::memcpy(reinterpret_cast<char*>(&address.sin_addr.s_addr),
+                    reinterpret_cast<char*>(server->h_addr),
+                    server->h_length);
+    }
+    else {
+        close(fd);
+        log<NUClear::ERROR>(fmt::format("Cannot resolve server name: {}", server_name));
+        return -1;
+    }
+
+    // Connect to the ip address
+    if (int error = connect(fd, (struct sockaddr*) &address, sizeof(address))) {
+        close(fd);
+        log<NUClear::ERROR>(fmt::format("Cannot connect server: {}. Error code {}", server_name, error));
+        return -1;
+    }
+
+    return fd;
+}
+
+Webots::Webots(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+
+    on<Configuration>("webots.yaml").then([this](const Configuration& cfg) {
+        // Use configuration here from file webots.yaml
+
+        // clang-format off
+        auto lvl = cfg["log_level"].as<std::string>();
+        if (lvl == "TRACE") { this->log_level = NUClear::TRACE; }
+        else if (lvl == "DEBUG") { this->log_level = NUClear::DEBUG; }
+        else if (lvl == "INFO") { this->log_level = NUClear::INFO; }
+        else if (lvl == "WARN") { this->log_level = NUClear::WARN; }
+        else if (lvl == "ERROR") { this->log_level = NUClear::ERROR; }
+        else if (lvl == "FATAL") { this->log_level = NUClear::FATAL; }
+        // clang-format on
+
+        int fd = tcpip_connect(cfg["server_address"].as<std::string>(), cfg["port"].as<int>());
+
+        // Tell webots who we are
+        int team_id  = cfg["team_id"].as<int>();
+        int robot_id = cfg["robot_id"].as<int>();
+        send_connect(fd, team_id, robot_id);
+
+        on<IO>(fd).then([this, fd]() {
+            // Receiving
+
+            // Get the size of the message
+            uint64_t N;
+            if (recv(fd, &N, sizeof(N), 0 != sizeof(N))) {
+                log<NUClear::ERROR>("Failed to read message size from TCP connection");
+                return;
+            }
+
+            // Get the message
+            std::vector<char> data(N, 0);
+            if (uint64_t(recv(fd, data.data(), N, 0)) != N) {
+                log<NUClear::ERROR>("Error: Failed to read message from TCP connection");
+                return;
+            }
+
+            // Deserialise the message into a neutron
+            SensorMeasurements msg = NUClear::util::serialise::Serialise<SensorMeasurements>::deserialise(data);
+
+            // Read each field of msg, translate it to our protobuf and emit the data
+            auto sensor_data = std::make_unique<DarwinSensors>();
+            sensor_data.FromSeconds(msg.time);
```

Comment:
According to the [API specs document](https://cdn.robocup.org/hl/wp/2021/03/v-hsc_simulator_api.pdf), the sensors measurement has no timestamp?

Reply:
And even if it did, `FromSeconds` is only valid on a protobuf timestamp message. You are using a `Neutron` here so `timestamp` is a `std::chrono::time_point`, so this should work I think
```suggestion
            sensor_data->timestamp = msg.time;
```

Reply:
Althought, perhaps we just want
```suggestion
            sensor_data->timestamp = NUClear::clock::now();
```